runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobInvocation.java

```diff
@@ -0,0 +1,41 @@
+package org.apache.beam.runners.fnexecution.jobsubmission;
+
+import io.grpc.stub.StreamObserver;
+import org.apache.beam.model.jobmanagement.v1.JobApi.JobMessage;
+import org.apache.beam.model.jobmanagement.v1.JobApi.JobState;
+
+/**
+ * Internal representation of a Job which has been invoked (prepared and run) by a client.
+ */
+public interface JobInvocation {
+
+  /**
+   * Start the job.
+   */
+  void start();
+
+  /**
+   * @return Unique identifier for the job invocation.
+   */
+  String getId();
+
+  /**
+   * Cancel the job.
+   */
+  void cancel();
+
+  /**
+   * Retrieve the job's current state.
+   */
+  JobState.Enum getState();
+
+  /**
+   * Observe job state changes with a {@link StreamObserver}.
+   */
+  void addStateObserver(StreamObserver<JobState.Enum> stateStreamObserver);
+
+  /**
+   * Observe job messages with a {@link StreamObserver}.
+   */
+  void addMessageObserver(StreamObserver<JobMessage> messageStreamObserver);
```

Comment:
Ditto on defining an interface like Consumer<JobMessage> instead of using StreamObserver

Reply:
You should use java exceptions to propagate this kind of information instead of exposing internal details. Like you wouldn't want the caller to call onCompleted to close the stream.

Reply:
Fair enough.  I'll just use a Consumer interface rather than defining something new then.

Reply:
fixed